File: mini-complete-example/src/main/scala/com/oreilly/learningsparkexamples/mini/scala/WordCount.scala
/**
 * Illustrates flatMap + countByValue for wordcount.
 */
package com.oreilly.learningsparkexamples.mini.scala

import org.apache.spark._
import org.apache.spark.SparkContext._

object WordCount {
    def main(args: Array[String]) {
      val inputFile = args(0)
      val outputFile = args(1)
      val conf = new SparkConf().setAppName("wordCount")
      val sc = new SparkContext(conf)
      // load our input data
      val input =  sc.textFile(inputFile)
      // split it up into words
      val words = input.flatMap(line => line.split(" "))
      // make word and count
      val counts = words.map(word => (word, 1)).reduceByKey{case (x, y) => x + y}
      // save the word count back out to a text file, causing evaluation
      counts.saveAsTextFile(outputFile)
    }
}

File: mini-complete-example/src/main/java/com/oreilly/learningsparkexamples/mini/java/WordCount.java
/**
 * Illustrates a wordcount in Java
 */
package com.oreilly.learningsparkexamples.mini.java;

import java.util.Arrays;
import java.util.List;
import java.lang.Iterable;

import scala.Tuple2;

import org.apache.commons.lang.StringUtils;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;


public class WordCount {
  public static void main(String[] args) throws Exception {
		String master = args[0];
    String inputFile = args[1];
    String outputFile = args[2];
    // Create a Java version of the Spark Context
		JavaSparkContext sc = new JavaSparkContext(
      master, "wordcount", System.getenv("SPARK_HOME"), System.getenv("JARS"));
    // load our input data
    JavaRDD<String> input = sc.textFile(inputFile);
    // split up into words
    JavaRDD<String> words = input.flatMap(
      new FlatMapFunction<String, String>() {
        public Iterable<String> call(String x) {
          return Arrays.asList(x.split(" "));
        }});
    // word and count
    JavaPairRDD<String, Integer> counts = words.mapToPair(
      new PairFunction<String, String, Integer>(){
        public Tuple2<String, Integer> call(String x){
          return new Tuple2(x, 1);
        }}).reduceByKey(new Function2<Integer, Integer, Integer>(){
            public Integer call(Integer x, Integer y){ return x + y;}});
    // save the word count back out to a text file, causing evaluation
    counts.saveAsTextFile(outputFile);
	}
}
